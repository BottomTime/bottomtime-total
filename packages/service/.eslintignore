coverage/
dist/

src/types/**/*.d.ts
