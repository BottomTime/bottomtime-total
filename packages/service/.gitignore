/assets/templates/
/docs/swagger.json
