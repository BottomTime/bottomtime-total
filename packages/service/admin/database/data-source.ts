import path from 'path';
import { DataSource } from 'typeorm';

export async function getDataSource(
  url: string,
  requireSsl: boolean,
): Promise<DataSource> {
  const client = new DataSource({
    type: 'postgres',
    url,
    entities: [path.resolve(__dirname, '../../src/data/**/*.entity.ts')],
    ssl: requireSsl,
  });
  return await client.initialize();
}
