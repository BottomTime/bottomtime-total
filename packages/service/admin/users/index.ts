import { CommandModule } from 'yargs';

import { createAdmin } from './create-admin';
import { getUserToken } from './get-user-token';

export const userModule: CommandModule<{
  'postgres-uri': string;
  'require-ssl': boolean;
}> = {
  command: 'user',

  describe: 'Commands for managing a user account',

  builder(yargs) {
    return yargs
      .command(
        'new-admin',
        'Creates a new admin user.',
        (yargs) => {
          return yargs
            .positional('username', {
              alias: 'u',
              description: "The user's username.",
              type: 'string',
            })
            .positional('password', {
              alias: 'p',
              description: 'The password to set on the new admin account',
              type: 'string',
            })
            .help();
        },
        async (yargs) => {
          await createAdmin({
            postgresRequireSsl: yargs.requireSsl,
            potgresUri: yargs.postgresUri,
            username: yargs.username,
            password: yargs.password,
          });
        },
      )
      .command(
        'token <user>',
        'Gets a token for a user',
        (yargs) => {
          return yargs
            .positional('user', {
              demandOption: true,
              description: "The user's username or email address.",
              type: 'string',
            })
            .help();
        },
        async (yargs) => {
          await getUserToken(yargs.postgresUri, yargs.requireSsl, yargs.user);
        },
      )
      .help();
  },

  async handler() {
    /* TODO */
  },
};
