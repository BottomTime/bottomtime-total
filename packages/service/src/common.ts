import { z } from 'zod';
import { DepthUnit } from './constants';

export const BooleanString = z
  .enum(['true', 'false'])
  .transform((value) => value === 'true');

export const DepthSchema = z.object({
  depth: z.number().min(0),
  unit: z.nativeEnum(DepthUnit),
});
export type Depth = z.infer<typeof DepthSchema>;

export const GpsCoordinatesSchema = z.object({
  lat: z.number().gte(-90).lte(90),
  lon: z.number().gte(-180).lte(180),
});
export type GpsCoordinates = z.infer<typeof GpsCoordinatesSchema>;

export interface Range {
  min: number;
  max: number;
}
