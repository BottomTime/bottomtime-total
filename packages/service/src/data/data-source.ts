import path from 'path';
import { DataSource, DataSourceOptions } from 'typeorm';

import { Config } from '../config';

export const PostgresDataSourceOptions: DataSourceOptions = {
  type: 'postgres',
  url: Config.postgresUri,
  entities: [path.join(__dirname, './**/*.entity.ts')],
  synchronize: false,
};

export async function initDataSource(): Promise<DataSource> {
  const AppDataSource = new DataSource(PostgresDataSourceOptions);
  return await AppDataSource.initialize();
}
