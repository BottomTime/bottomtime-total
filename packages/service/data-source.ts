import path from 'path';
import { DataSource } from 'typeorm';

import { PostgresDataSourceOptions } from './src/data/data-source';

export default new DataSource({
  ...PostgresDataSourceOptions,
  migrations: [path.join(__dirname, './migrations/*.ts')],
});
