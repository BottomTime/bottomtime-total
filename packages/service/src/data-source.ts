import 'reflect-metadata';
import { DataSourceOptions } from 'typeorm';

import { Config } from './config';
import * as Entities from './data';

export const PostgresDataSourceOptions: DataSourceOptions = {
  type: 'postgres',
  url: Config.postgresUri,
  entities: Object.values(Entities),
  synchronize: false,
};
