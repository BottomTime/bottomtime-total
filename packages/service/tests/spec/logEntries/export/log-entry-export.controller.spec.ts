import { UserRole } from '@bottomtime/api';

import { HttpServer, INestApplication } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';

import {
  LogEntryAirEntity,
  LogEntryEntity,
  LogEntrySampleEntity,
  UserEntity,
} from 'src/data';
import { DiveSitesModule } from 'src/diveSites';
import { LogEntryFactory } from 'src/logEntries';
import { LogEntryExportController } from 'src/logEntries/export/log-entry-export.controller';
import { LogEntryExportService } from 'src/logEntries/export/log-entry-export.service';
import { OperatorsModule } from 'src/operators';
import { UsersModule } from 'src/users';
import request from 'supertest';
import { dataSource } from 'tests/data-source';
import {
  createAuthHeader,
  createTestApp,
  createTestLogEntry,
  createTestUser,
} from 'tests/utils';
import { Repository } from 'typeorm';

const Username = 'ExportyMcExportface_theThird';

function getUrl(username = Username): string {
  return `/api/users/${username}/logbook/export/json`;
}

describe('Log Entry Export E2E tests', () => {
  let app: INestApplication;
  let server: HttpServer;

  let Users: Repository<UserEntity>;
  let Entries: Repository<LogEntryEntity>;
  let Air: Repository<LogEntryAirEntity>;

  let owner: UserEntity;
  let otherUser: UserEntity;
  let admin: UserEntity;

  let entries: LogEntryEntity[];
  let air: LogEntryAirEntity[];

  let authHeader: [string, string];
  let otherAuthHeader: [string, string];
  let adminAuthHeader: [string, string];

  beforeAll(async () => {
    Users = dataSource.getRepository(UserEntity);
    Entries = dataSource.getRepository(LogEntryEntity);
    Air = dataSource.getRepository(LogEntryAirEntity);

    app = await createTestApp({
      imports: [
        TypeOrmModule.forFeature([
          LogEntryEntity,
          LogEntryAirEntity,
          LogEntrySampleEntity,
        ]),
        UsersModule,
        DiveSitesModule,
        OperatorsModule,
      ],
      providers: [LogEntryExportService, LogEntryFactory],
      controllers: [LogEntryExportController],
    });

    owner = createTestUser({
      username: Username,
      usernameLowered: Username.toLowerCase(),
    });
    otherUser = createTestUser();
    admin = createTestUser({ role: UserRole.Admin });

    entries = Array.from({ length: 450 }, () => createTestLogEntry(owner)).sort(
      (a, b) => a.entryTime.valueOf() - b.entryTime.valueOf(),
    );
    air = entries.reduce<LogEntryAirEntity[]>((acc, entry) => {
      if (entry.air) acc.push(...entry.air);
      return acc;
    }, []);

    [authHeader, otherAuthHeader, adminAuthHeader] = await Promise.all([
      createAuthHeader(owner.id),
      createAuthHeader(otherUser.id),
      createAuthHeader(admin.id),
    ]);

    server = app.getHttpServer();
  });

  beforeEach(async () => {
    await Users.save([owner, otherUser, admin]);
    await Entries.save(entries);
    await Air.save(air);
  });

  afterAll(async () => {
    await app.close();
  });

  it('will export the indicated logbook, unfiltered', async () => {
    const { body } = await request(server)
      .get(getUrl())
      .set(...authHeader)
      .expect(200);
    expect(body.data.length).toBe(entries.length);
    expect(body.totalCount).toBe(entries.length);
  });

  it('will export with a few filters', async () => {});

  it('will export with an alternate sort order', async () => {});

  it('will allow specific entries to be omitted', async () => {});

  it('will allow specific entries to be included', async () => {});

  it('will return a 400 response if the query string is invalid', async () => {
    const { body } = await request(server)
      .get(getUrl())
      .set(...authHeader)
      .query({
        query: true,
        startDate: 'yesterday',
        endDate: 'tomorrow',
        location: {
          lat: 1111,
          lon: -1883,
        },
        radius: -8,
        minRating: 'ok',
        maxRating: 5.5,
        sortBy: 'time',
        sortOrder: 'backward',
        include: 7,
        omit: 'sure',
      })
      .expect(400);

    expect(body.details).toMatchSnapshot();
  });

  it('will return a 401 response if the request is unauthenticated', async () => {
    await request(server).get(getUrl()).expect(401);
  });

  it('will return a 403 response if the user is not authorized to export the logbook', async () => {
    await request(server)
      .get(getUrl())
      .set(...otherAuthHeader)
      .expect(403);
  });

  it('will return a 404 response if the indicated logbook does not exist', async () => {
    await request(server)
      .get(getUrl('no_such_user'))
      .set(...adminAuthHeader)
      .expect(404);
  });
});
