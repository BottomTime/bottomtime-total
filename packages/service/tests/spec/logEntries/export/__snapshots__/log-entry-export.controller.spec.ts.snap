// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Log Entry Export E2E tests will return a 400 response if the query string is invalid 1`] = `
{
  "issues": [
    {
      "code": "invalid_type",
      "expected": "number",
      "message": "Expected number, received nan",
      "path": [
        "startDate",
      ],
      "received": "nan",
    },
    {
      "code": "invalid_type",
      "expected": "number",
      "message": "Expected number, received nan",
      "path": [
        "endDate",
      ],
      "received": "nan",
    },
    {
      "code": "invalid_type",
      "expected": "string",
      "message": "Expected string, received object",
      "path": [
        "location",
      ],
      "received": "object",
    },
    {
      "code": "too_small",
      "exact": false,
      "inclusive": false,
      "message": "Number must be greater than 0",
      "minimum": 0,
      "path": [
        "radius",
      ],
      "type": "number",
    },
    {
      "code": "invalid_type",
      "expected": "number",
      "message": "Expected number, received nan",
      "path": [
        "minRating",
      ],
      "received": "nan",
    },
    {
      "code": "too_big",
      "exact": false,
      "inclusive": true,
      "maximum": 5,
      "message": "Number must be less than or equal to 5",
      "path": [
        "maxRating",
      ],
      "type": "number",
    },
    {
      "code": "invalid_enum_value",
      "message": "Invalid enum value. Expected 'entryTime' | 'logNumber' | 'rating', received 'time'",
      "options": [
        "entryTime",
        "logNumber",
        "rating",
      ],
      "path": [
        "sortBy",
      ],
      "received": "time",
    },
    {
      "code": "invalid_enum_value",
      "message": "Invalid enum value. Expected 'asc' | 'desc', received 'backward'",
      "options": [
        "asc",
        "desc",
      ],
      "path": [
        "sortOrder",
      ],
      "received": "backward",
    },
    {
      "code": "invalid_type",
      "expected": "array",
      "message": "Expected array, received string",
      "path": [
        "include",
      ],
      "received": "string",
    },
    {
      "code": "invalid_type",
      "expected": "array",
      "message": "Expected array, received string",
      "path": [
        "omit",
      ],
      "received": "string",
    },
  ],
}
`;
